tests/smoke/alp_zgeqrf.cpp

```diff
@@ -0,0 +1,301 @@
+/*
+ *   Copyright 2021 Huawei Technologies Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#ifdef _COMPLEX
+#include <complex>
+#include <cmath>
+#include <iomanip>
+#endif
+
+#include <alp.hpp>
+#include <alp/algorithms/householder_qr.hpp>
+#include <graphblas/utils/iscomplex.hpp> // use from grb
+#include "../utils/print_alp_containers.hpp"
+
+//once TEMPDISABLE is removed the code should be in the final version
+#define TEMPDISABLE
+
+using namespace alp;
+
+using BaseScalarType = double;
+using Orthogonal = structures::Orthogonal;
+using General = structures::General;
+
+#ifdef _COMPLEX
+using ScalarType = std::complex< BaseScalarType >;
+#else
+using ScalarType = BaseScalarType;
+#endif
+
+constexpr BaseScalarType tol = 1.e-10;
+constexpr size_t RNDSEED = 1;
+
+//** generate random rectangular matrix data: complex version */
+template<
+	typename T
+>
+std::vector< T > generate_rectangular_matrix_data(
+	size_t N,
+	size_t M,
+	const typename std::enable_if<
+		grb::utils::is_complex< T >::value,
+		void
+	>::type * const = nullptr
+) {
+	std::vector< T > data( N * M );
+	std::srand( RNDSEED );
```

Comment:
Due to this line, the function `generate_rectangular_matrix_data` will always produce the same data, no matter how many times it is called in an application. While in this one it is called only once, the side-effect should be noted for future use.
Maybe it could be good to move the call to `srand` outside, so it is called only once (again - not relevant for this particular test, but rather for potential future expansion).